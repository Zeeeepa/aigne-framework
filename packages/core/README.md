# @aigne/core

[![GitHub star chart](https://img.shields.io/github/stars/AIGNE-io/aigne-framework?style=flat-square)](https://star-history.com/#AIGNE-io/aigne-framework)
[![Open Issues](https://img.shields.io/github/issues-raw/AIGNE-io/aigne-framework?style=flat-square)](https://github.com/AIGNE-io/aigne-framework/issues)
[![codecov](https://codecov.io/gh/AIGNE-io/aigne-framework/graph/badge.svg?token=DO07834RQL)](https://codecov.io/gh/AIGNE-io/aigne-framework)
[![NPM Version](https://img.shields.io/npm/v/@aigne/core)](https://www.npmjs.com/package/@aigne/core)
[![Elastic-2.0 licensed](https://img.shields.io/npm/l/@aigne/core)](https://github.com/AIGNE-io/aigne-framework/blob/main/LICENSE)

**English** | [中文](README.zh.md)

Core library of [AIGNE Framework](https://github.com/AIGNE-io/aigne-framework) for building AI-powered applications.

## Introduction

`@aigne/core` is the foundation component of [AIGNE Framework](https://github.com/AIGNE-io/aigne-framework), providing the essential modules and tools needed to build AI-driven applications. This package implements the core functionalities of the framework, including agent systems, aigne environment, model integrations, and workflow pattern support.

## Features

- **Multiple AI Model Support**: Built-in support for OpenAI, Gemini, Claude, Nova, and other mainstream AI models, easily extensible to support additional models
- **Agent System**: Powerful agent abstractions supporting AI agents, function agents, MCP agents, and more
- **AIGNE Environment**: Flexible handling communication between agents and workflow execution
- **Workflow Patterns**: Support for sequential, concurrent, routing, handoff, and other workflow patterns
- **MCP Protocol Integration**: Seamless integration with external systems through the Model Context Protocol
- **TypeScript Support**: Comprehensive type definitions providing an excellent development experience

## Installation

```bash
# Using npm
npm install @aigne/core

# Using yarn
yarn add @aigne/core

# Using pnpm
pnpm add @aigne/core
```

## Basic Usage

```typescript
import { AIAgent, AIGNE } from "@aigne/core";
import { OpenAIChatModel } from "@aigne/core/models/openai-chat-model.js";

// Create AI model instance
const model = new OpenAIChatModel({
  apiKey: process.env.OPENAI_API_KEY,
  model: process.env.DEFAULT_CHAT_MODEL || "gpt-4-turbo",
});

// Create AI agent
const agent = AIAgent.from({
  name: "Assistant",
  instructions: "You are a helpful assistant.",
});

// Create AIGNE
const aigne = new AIGNE({ model });

// Use the AIGNE to invoke the agent
const userAgent = await aigne.invoke(agent);

// Send a message to the agent
const response = await userAgent.invoke("Hello, can you help me write a short article?");
console.log(response);
```

## Module Structure

- `agents/`: Agent implementations, including AI agents, function agents, MCP agents, etc.
- `aigne/`: AIGNE is agents execution environment
- `loader/`: Loader-related functionality
- `models/`: Integration with various AI models
- `prompt/`: Prompt handling functionality
- `utils/`: Utility functions and helper methods

## Documentation

For more detailed API documentation, please refer to:

- [Agent API](../../docs/apis/agent-api.md)
- [AI Agent API](../../docs/apis/ai-agent-api.md)
- [Function Agent API](../../docs/apis/function-agent-api.md)
- [MCP Agent API](../../docs/apis/mcp-agent-api.md)
- [AIGNE API](../../docs/apis/aigne-api.md)
- [Server/Client API](../../docs/apis/server-client-api.md)

## License

Elastic-2.0
