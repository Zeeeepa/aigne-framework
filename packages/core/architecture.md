## Class Definitions

```mermaid
classDiagram
    class PromptBuilderBuildOptions {
        +Context context
        +Agent agent
        +object input
        +ChatModel model
    }

    class Prompt {
        +List~object~ messages
        +List~Agent~ skills
        +object toolChoice
        +object responseFormat
    }

    PromptBuilder ..> PromptBuilderBuildOptions: dependency
    PromptBuilder ..> Prompt: dependency
    class PromptBuilder {
        +build(PromptBuilderBuildOptions options) Prompt
    }

    ChatModel --|> Agent: inheritance
    class ChatModel {
    }

    ImageModel --|> Agent: inheritance
    class ImageModel {
    }

    class Agent {
        +string name
        +string description
        +object inputSchema
        +object outputSchema
        +List~string~ subscribeTopic
        +List~string~ publishTopic
        +List~Agent~ skills
        +MemoryAgent memory

        +invoke(object input, Context context) object
        +shutdown() void
        +process(object input, Context context) object*
        -preprocess() void
        -postprocess() void
    }

    AIAgent --|> Agent: inheritance
    AIAgent *.. PromptBuilder: composition
    class AIAgent {
        +ChatModel model
        +PromptBuilder instructions
        +string outputKey
        +object toolChoice
        +boolean catchToolsError
    }

    TeamAgent --|> Agent: inheritance
    class TeamAgent {
        +ProcessMode mode
    }

    ImageAgent --|> Agent: inheritance
    ImageAgent *.. PromptBuilder: composition
    ImageAgent *.. ImageModel: composition
    class ImageAgent {
        +ImageModel model
        +PromptBuilder instructions
    }

    FunctionAgent --|> Agent: inheritance
    class FunctionAgent {
        +Function process
    }

    RPCAgent --|> Agent: inheritance
    class RPCAgent {
        +string url
    }

    MCPAgent --|> Agent: inheritance
    MCPAgent *.. MCPClient: composition
    class MCPAgent {
        +MCPClient client
    }

    class MCPClient {
    }

    class Message {
        +object output
    }

    MessageQueue ..> Message: dependency
    class MessageQueue {
        +publish(string topic, Message message) void
        +subscribe(string topic, Function callback) void
        +unsubscribe(string topic, Function callback) void
    }

    Context *.. ChatModel: composition
    Context *.. ImageModel: composition
    Context --|> MessageQueue: inheritance
    class Context {
        +ChatModel model
        +ImageModel imageModel
        +List~Agent~ skills

        +invoke(Agent agent, object input) object
        +publish(string topic, Message message) void
        +subscribe(string topic, Function callback) void
        +unsubscribe(string topic, Function callback) void
    }

    UserAgent --|> Agent: inheritance
    class UserAgent {
    }

    class EventEmitter {
        +on(): void
        +emit(): void
    }

    AIGNE --|> EventEmitter: inheritance
    class AIGNE {
        +string name
        +string description
        +ChatModel model
        +object limits
        +List~Agent~ agents
        +List~Agent~ skills

        +load(string path) AIGNE$
        +addAgent(Agent agent) void
        +invoke(Agent agent) UserAgent
        +invoke(Agent agent, string input) object
        +invoke(Agent agent, object input) object
        +publish(string topic, Message message) void
        +subscribe(string topic, Function callback) void
        +unsubscribe(string topic, Function callback) void
        +shutdown() void
    }
```
