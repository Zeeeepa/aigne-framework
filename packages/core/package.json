{
  "name": "@aigne/core",
  "version": "1.20.1",
  "description": "AIGNE core library for building AI-powered applications",
  "publishConfig": {
    "access": "public"
  },
  "author": "Arcblock <blocklet@arcblock.io> https://github.com/blocklet",
  "homepage": "https://github.com/AIGNE-io/aigne-framework",
  "license": "Elastic-2.0",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/AIGNE-io/aigne-framework"
  },
  "files": [
    "lib/cjs",
    "lib/dts",
    "lib/esm",
    "LICENSE",
    "README.md",
    "CHANGELOG.md"
  ],
  "type": "module",
  "main": "./lib/cjs/index.js",
  "module": "./lib/esm/index.js",
  "types": "./lib/dts/index.d.ts",
  "exports": {
    ".": {
      "import": "./lib/esm/index.js",
      "require": "./lib/cjs/index.js",
      "types": "./lib/dts/index.d.ts"
    },
    "./*": {
      "import": "./lib/esm/*",
      "require": "./lib/cjs/*",
      "types": "./lib/dts/*"
    }
  },
  "typesVersions": {
    "*": {
      ".": [
        "./lib/dts/index.d.ts"
      ],
      "agents/*": [
        "./lib/dts/agents/*"
      ],
      "aigne/*": [
        "./lib/dts/aigne/*"
      ],
      "loader/*": [
        "./lib/dts/loader/*"
      ],
      "memory/*": [
        "./lib/dts/memory/*"
      ],
      "prompt/*": [
        "./lib/dts/prompt/*"
      ],
      "utils/*": [
        "./lib/dts/utils/*"
      ]
    }
  },
  "scripts": {
    "lint": "tsc --noEmit",
    "build": "tsc --build scripts/tsconfig.build.json",
    "clean": "rimraf lib test/coverage",
    "prepublishOnly": "run-s clean build",
    "test": "bun --cwd test test",
    "test:coverage": "bun --cwd test test --coverage --coverage-reporter=lcov --coverage-reporter=text",
    "postbuild": "echo '{\"type\": \"module\"}' > lib/esm/package.json && echo '{\"type\": \"commonjs\"}' > lib/cjs/package.json"
  },
  "dependencies": {
    "@aigne/json-schema-to-zod": "^1.3.3",
    "@aigne/platform-helpers": "workspace:^",
    "@modelcontextprotocol/sdk": "^1.11.0",
    "@types/debug": "^4.1.12",
    "camelize-ts": "^3.0.0",
    "content-type": "^1.0.5",
    "debug": "^4.4.0",
    "eventsource-parser": "^3.0.1",
    "fast-deep-equal": "^3.1.3",
    "immer": "^10.1.1",
    "mustache": "^4.2.0",
    "nanoid": "^5.1.5",
    "p-retry": "^6.2.1",
    "raw-body": "^3.0.0",
    "strict-event-emitter": "^0.5.1",
    "ufo": "^1.6.1",
    "uuid": "^11.1.0",
    "yaml": "^2.7.1",
    "zod": "^3.24.4",
    "zod-to-json-schema": "^3.24.5"
  },
  "devDependencies": {
    "@types/bun": "^1.2.12",
    "@types/compression": "^1.7.5",
    "@types/content-type": "^1.1.8",
    "@types/express": "^5.0.1",
    "@types/mustache": "^4.2.6",
    "@types/node": "^22.15.15",
    "compression": "^1.8.0",
    "detect-port": "^2.1.0",
    "express": "^5.1.0",
    "npm-run-all": "^4.1.5",
    "rimraf": "^6.0.1",
    "typescript": "^5.8.3"
  }
}
