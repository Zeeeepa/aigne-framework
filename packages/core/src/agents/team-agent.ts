import type { Context } from "../aigne/context.js";
import { mergeAgentResponseChunk, readableStreamToAsyncIterator } from "../utils/stream-utils.js";
import { type PromiseOrValue, isEmpty } from "../utils/type-utils.js";
import {
  Agent,
  type AgentOptions,
  type AgentProcessResult,
  type AgentResponseChunk,
  type Message,
} from "./agent.js";

/**
 * Defines the processing modes available for a TeamAgent.
 *
 * The processing mode determines how the agents within a team are executed
 * and how their outputs are combined.
 */
export enum ProcessMode {
  /**
   * Process the agents one by one, passing the output of each agent to the next.
   *
   * In sequential mode, agents execute in order, with each agent receiving the
   * combined output from all previous agents as part of its input.
   */
  sequential = "sequential",

  /**
   * Process all agents in parallel, merging the output of all agents.
   *
   * In parallel mode, all agents execute simultaneously, each receiving the same
   * initial input. Their outputs are then combined based on output key ownership.
   */
  parallel = "parallel",
}

/**
 * Configuration options for creating a TeamAgent.
 *
 * These options extend the base AgentOptions and add team-specific settings.
 */
export interface TeamAgentOptions<I extends Message, O extends Message> extends AgentOptions<I, O> {
  /**
   * The method to process the agents in the team.
   * @default {ProcessMode.sequential}
   */
  mode?: ProcessMode;
}

/**
 * TeamAgent coordinates a group of agents working together to accomplish tasks.
 *
 * A TeamAgent manages a collection of agents (its skills) and orchestrates their
 * execution according to a specified processing mode. It provides mechanisms for
 * agents to work either sequentially (one after another) or in parallel (all at once),
 * with appropriate handling of their outputs.
 *
 * TeamAgent is particularly useful for:
 * - Creating agent workflows where output from one agent feeds into another
 * - Executing multiple agents simultaneously and combining their results
 * - Building complex agent systems with specialized components working together
 *
 * @example
 * Here's an example of creating a sequential TeamAgent:
 * {@includeCode ../../test/agents/team-agent.test.ts#example-team-agent-sequential}
 */
export class TeamAgent<I extends Message, O extends Message> extends Agent<I, O> {
  /**
   * Create a TeamAgent from the provided options.
   *
   * @param options Configuration options for the TeamAgent
   * @returns A new TeamAgent instance
   *
   * @example
   * Here's an example of creating a sequential TeamAgent:
   * {@includeCode ../../test/agents/team-agent.test.ts#example-team-agent-sequential}
   *
   * @example
   * Here's an example of creating a parallel TeamAgent:
   * {@includeCode ../../test/agents/team-agent.test.ts#example-team-agent-parallel}
   */
  static from<I extends Message, O extends Message>(options: TeamAgentOptions<I, O>) {
    return new TeamAgent(options);
  }

  /**
   * Create a new TeamAgent instance.
   *
   * @param options Configuration options for the TeamAgent
   */
  constructor(options: TeamAgentOptions<I, O>) {
    super(options);
    this.mode = options.mode ?? ProcessMode.sequential;
  }

  /**
   * The processing mode that determines how agents in the team are executed.
   *
   * This can be either sequential (one after another) or parallel (all at once).
   */
  mode: ProcessMode;

  /**
   * Process an input message by routing it through the team's agents.
   *
   * Depending on the team's processing mode, this will either:
   * - In sequential mode: Pass input through each agent in sequence, with each agent
   *   receiving the combined output from previous agents
   * - In parallel mode: Process input through all agents simultaneously and combine their outputs
   *
   * @param input The message to process
   * @param context The execution context
   * @returns A stream of message chunks that collectively form the response
   */
  process(input: I, context: Context): PromiseOrValue<AgentProcessResult<O>> {
    switch (this.mode) {
      case ProcessMode.sequential:
        return this._processSequential(input, context);
      case ProcessMode.parallel:
        return this._processParallel(input, context);
    }
  }

  /**
   * Process input sequentially through each agent in the team.
   *
   * This method:
   * 1. Executes each agent in order
   * 2. Passes the combined output from previous agents to the next agent
   * 3. Yields output chunks as they become available
   * 4. Updates the team's agent list with any changes that occurred during processing
   *
   * @param input The message to process
   * @param context The execution context
   * @returns A stream of message chunks from all agents
   *
   * @private
   */
  async *_processSequential(input: I, context: Context): PromiseOrValue<AgentProcessResult<O>> {
    const output: Message = {};

    // Clone the agents to run, so that we can update the agents list during the loop
    const agents = [...this.skills];
    const newAgents: Agent[] = [];

    for (const agent of agents) {
      const [o, transferToAgent] = await context.invoke(
        agent,
        { ...input, ...output },
        { returnActiveAgent: true, streaming: true },
      );

      for await (const chunk of readableStreamToAsyncIterator(o)) {
        yield chunk;
        mergeAgentResponseChunk(output, chunk);
      }
      newAgents.push(await transferToAgent);
    }

    this.skills.splice(0);
    this.skills.push(...newAgents);
  }

  /**
   * Process input in parallel through all agents in the team.
   *
   * This method:
   * 1. Executes all agents simultaneously with the same input
   * 2. Yields combined output chunks
   * 3. Updates the team's agent list with any changes that occurred during processing
   *
   * @param input The message to process
   * @param context The execution context
   * @returns A stream of combined message chunks from all agents
   *
   * @private
   */
  async *_processParallel(input: I, context: Context): PromiseOrValue<AgentProcessResult<O>> {
    const result = await Promise.all(
      this.skills.map((agent) =>
        context.invoke(agent, input, { returnActiveAgent: true, streaming: true }),
      ),
    );

    const streams = result.map((i) => i[0]);

    type Reader = ReadableStreamDefaultReader<AgentResponseChunk<Message>>;
    type ReaderResult = ReadableStreamReadResult<AgentResponseChunk<Message>>;

    type Task = Promise<{ index: number; reader: Reader } & ReaderResult>;

    const read = async (index: number, reader: Reader): Task => {
      const promise = reader.read();
      return promise.then((result) => ({ ...result, reader, index }));
    };

    const tasks = new Map(streams.map((stream, index) => [index, read(index, stream.getReader())]));

    // NOTE: Flag to check if the output key is used by agent at the index,
    const outputKeyUsed = new Map<string, number>();

    while (tasks.size) {
      const { value, done, reader, index } = await Promise.race(tasks.values());
      tasks.delete(index);
      if (!done) {
        tasks.set(index, read(index, reader));
      }

      if (value) {
        let {
          delta: { text, ...delta },
        } = value;

        if (text) {
          for (const key of Object.keys(text)) {
            // the output key is unused, add to map to lock it
            if (!outputKeyUsed.has(key)) {
              outputKeyUsed.set(key, index);
            }
            // the output key is used by the agent at the index, abandon it
            else if (outputKeyUsed.get(key) !== index) {
              delete text[key];
            }
          }
          if (isEmpty(text)) {
            text = undefined;
          }
        }

        if (!isEmpty(delta.json) || !isEmpty(text)) yield { delta: { ...delta, text } };
      }
    }

    const agents = await Promise.all(result.map((i) => i[1]));
    this.skills.splice(0);
    this.skills.push(...agents);
  }
}
