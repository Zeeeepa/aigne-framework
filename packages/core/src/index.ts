export * from "./agents/agent.js";
export * from "./agents/ai-agent.js";
export * from "./agents/chat-model.js";
export * from "./agents/guide-rail-agent.js";
export * from "./agents/mcp-agent.js";
export * from "./agents/team-agent.js";
export * from "./agents/types.js";
export * from "./agents/user-agent.js";
export * from "./aigne/index.js";
export * from "./memory/index.js";
export * from "./prompt/prompt-builder.js";
export * from "./prompt/template.js";
export * from "./utils/stream-utils.js";
