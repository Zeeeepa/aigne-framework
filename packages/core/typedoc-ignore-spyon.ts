import type { Application, Context, Reflection } from "typedoc";
import {
  EmitHint,
  NewLineKind,
  type Node,
  ScriptKind,
  ScriptTarget,
  type Statement,
  createPrinter,
  createSourceFile,
  factory,
  isCallExpression,
} from "typescript";

const IGNORE_EXPRESSIONS = ["spyOn"];

export function load(app: Application) {
  // @ts-ignore can not resolve `on` method of `converter` in `typedoc`
  app.converter.on("resolveReflection", (_: Context, reflection: Reflection) => {
    if (!reflection.comment) return;

    const exampleTags = reflection.comment.getTags("@example");

    for (const tag of exampleTags) {
      for (const content of tag.content) {
        if (content.kind === "code") {
          const code = content.text
            .trim()
            .replace(/^\s*```\S+/, "")
            .replace(/```$/, "");

          const c = removeFunctionCallsFromCode(code, IGNORE_EXPRESSIONS);
          content.text = `\`\`\`ts\n${c}\n\`\`\``;
        }
      }
    }
  });
}

function removeFunctionCallsFromCode(sourceCode: string, functionName: string[]): string {
  const sourceFile = createSourceFile(
    "temp.ts",
    sourceCode,
    ScriptTarget.Latest,
    true,
    ScriptKind.TS,
  );

  const printer = createPrinter({
    newLine: NewLineKind.LineFeed,
  });
  const statements: Statement[] = [];

  sourceFile.forEachChild((node) => {
    let shouldRemove = false;

    function check(node: Node) {
      if (isCallExpression(node)) {
        const expressionText = node.expression.getText();
        if (functionName.includes(expressionText)) {
          shouldRemove = true;
        }
      }
      node.forEachChild(check);
    }

    check(node);

    if (!shouldRemove) {
      statements.push(node as Statement);
    }
  });

  const newSourceFile = factory.updateSourceFile(sourceFile, statements);

  return statements
    .map((statement) => printer.printNode(EmitHint.Unspecified, statement, newSourceFile))
    .join("\n\n")
    .trim();
}
