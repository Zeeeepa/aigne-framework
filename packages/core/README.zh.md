# @aigne/core

[![GitHub star chart](https://img.shields.io/github/stars/AIGNE-io/aigne-framework?style=flat-square)](https://star-history.com/#AIGNE-io/aigne-framework)
[![Open Issues](https://img.shields.io/github/issues-raw/AIGNE-io/aigne-framework?style=flat-square)](https://github.com/AIGNE-io/aigne-framework/issues)
[![codecov](https://codecov.io/gh/AIGNE-io/aigne-framework/graph/badge.svg?token=DO07834RQL)](https://codecov.io/gh/AIGNE-io/aigne-framework)
[![NPM Version](https://img.shields.io/npm/v/@aigne/core)](https://www.npmjs.com/package/@aigne/core)
[![Elastic-2.0 licensed](https://img.shields.io/npm/l/@aigne/core)](https://github.com/AIGNE-io/aigne-framework/blob/main/LICENSE)

[English](README.md) | **中文**

[AIGNE 框架](https://github.com/AIGNE-io/aigne-framework)的核心库，用于构建 AI 驱动的应用程序。

## 简介

`@aigne/core` 是 [AIGNE 框架](https://github.com/AIGNE-io/aigne-framework)的基础核心组件，提供了构建 AI 驱动应用程序所需的基本模块和工具。该包实现了框架的核心功能，包括代理系统、执行引擎、模型集成以及工作流模式支持等。

## 特性

- **多种 AI 模型支持**：内置支持 OpenAI、Gemini、Claude、Nova 等主流 AI 模型，可轻松扩展支持其他模型
- **代理系统**：强大的代理抽象，支持 AI 代理、函数代理、MCP 代理等多种类型
- **执行引擎**：灵活的执行引擎，处理代理之间的通信和工作流执行
- **工作流模式**：支持顺序、并发、路由、交接等多种工作流模式
- **MCP 协议集成**：通过模型上下文协议（Model Context Protocol）实现与外部系统的无缝集成
- **TypeScript 支持**：完善的类型定义，提供良好的开发体验

## 安装

```bash
# 使用 npm
npm install @aigne/core

# 使用 yarn
yarn add @aigne/core

# 使用 pnpm
pnpm add @aigne/core
```

## 基本用法

```typescript
import { AIAgent, AIGNE } from "@aigne/core";
import { OpenAIChatModel } from "@aigne/core/models/openai-chat-model.js";

// 创建 AI 模型实例
const model = new OpenAIChatModel({
  apiKey: process.env.OPENAI_API_KEY,
  model: process.env.DEFAULT_CHAT_MODEL || "gpt-4-turbo",
});

// 创建 AI 代理
const agent = AIAgent.from({
  name: "助手",
  instructions: "你是一个有帮助的助手。",
});

// 创建执行引擎
const aigne = new AIGNE({ model });

// 使用执行引擎调用代理
const userAgent = await aigne.invoke(agent);

// 向代理发送消息
const response = await userAgent.invoke("你好，能帮我写一篇短文吗？");
console.log(response);
```

## 模块结构

- `agents/`: 代理相关的实现，包括 AI 代理、函数代理、MCP 代理等
- `aigne/`: AIGNE 提供 Agents 执行环境
- `loader/`: 加载器相关功能
- `models/`: 各种 AI 模型的集成
- `prompt/`: 提示词处理相关功能
- `utils/`: 工具函数和辅助方法

## 文档

更多详细的 API 文档，请参考：

- [Agent API](../../docs/apis/agent-api.zh.md)
- [AI Agent API](../../docs/apis/ai-agent-api.zh.md)
- [Function Agent API](../../docs/apis/function-agent-api.zh.md)
- [MCP Agent API](../../docs/apis/mcp-agent-api.zh.md)
- [AIGNE API](../../docs/apis/aigne-api.zh.md)
- [服务器/客户端 API](../../docs/apis/server-client-api.zh.md)

## 协议

Elastic-2.0
