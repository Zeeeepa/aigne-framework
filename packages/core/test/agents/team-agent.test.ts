import { expect, spyOn, test } from "bun:test";
import assert from "node:assert";
import { AIAgent, AIGNE, FunctionAgent, type Message } from "@aigne/core";
import { ProcessMode, TeamAgent } from "@aigne/core/agents/team-agent.js";
import {
  readableStreamToArray,
  stringToAgentResponseStream,
} from "@aigne/core/utils/stream-utils.js";
import { OpenAIChatModel } from "../_mocks/mock-models.js";

test("TeamAgent.from with sequential mode", async () => {
  // #region example-team-agent-sequential

  // Create individual specialized agents
  const translatorAgent = FunctionAgent.from({
    name: "translator",
    process: (input: Message) => ({
      translation: `${input.text} (translation)`,
    }),
  });

  const formatterAgent = FunctionAgent.from({
    name: "formatter",
    process: (input: Message) => ({
      formatted: `[formatted] ${input.translation || input.text}`,
    }),
  });

  // Create a sequential TeamAgent with specialized agents
  const teamAgent = TeamAgent.from({
    name: "sequential-team",
    mode: ProcessMode.sequential,
    skills: [translatorAgent, formatterAgent],
  });

  const result = await teamAgent.invoke({ text: "Hello world" });

  expect(result).toEqual({
    translation: "Hello world (translation)",
    formatted: "[formatted] Hello world (translation)",
  });
  console.log(result);
  // Expected output: {
  //   translation: "Hello world (translation)",
  //   formatted: "[formatted] Hello world (translation)"
  // }

  // #endregion example-team-agent-sequential
});

test("TeamAgent.from with parallel mode", async () => {
  // #region example-team-agent-parallel

  const googleSearch = FunctionAgent.from({
    name: "google-search",
    process: (input: Message) => ({
      googleResults: `Google search results for ${input.query}`,
    }),
  });

  const braveSearch = FunctionAgent.from({
    name: "brave-search",
    process: (input: Message) => ({
      braveResults: `Brave search results for ${input.query}`,
    }),
  });

  const teamAgent = TeamAgent.from({
    name: "parallel-team",
    mode: ProcessMode.parallel,
    skills: [googleSearch, braveSearch],
  });

  const result = await teamAgent.invoke({ query: "AI news" });

  expect(result).toEqual({
    googleResults: "Google search results for AI news",
    braveResults: "Brave search results for AI news",
  });

  console.log(result);
  // Expected output: {
  //   googleResults: "Google search results for AI news",
  //   braveResults: "Brave search results for AI news"
  // }

  // #endregion example-team-agent-parallel
});

const processModes = Object.values(ProcessMode);

test.each(processModes)(
  "TeamAgent should return streaming response with %s process method (multiple agent with different output keys)",
  async (mode) => {
    const model = new OpenAIChatModel();

    const aigne = new AIGNE({ model });

    const first = AIAgent.from({
      outputKey: "first",
      inputKey: "message",
    });

    const second = AIAgent.from({
      outputKey: "second",
      inputKey: "message",
    });

    spyOn(model, "process")
      .mockReturnValueOnce(Promise.resolve(stringToAgentResponseStream("Hello, ")))
      .mockReturnValueOnce(Promise.resolve(stringToAgentResponseStream("Hello, world!")));

    const team = TeamAgent.from({
      skills: [first, second],
      mode,
    });

    const stream = await aigne.invoke(team, { message: "hello" }, { streaming: true });

    expect(readableStreamToArray(stream)).resolves.toMatchSnapshot();
  },
);

test.each(processModes)(
  "TeamAgent should return streaming response with %s process method (multiple agent with same output key)",
  async (mode) => {
    const model = new OpenAIChatModel();

    const aigne = new AIGNE({ model });

    const first = AIAgent.from({
      outputKey: "text",
    });

    const second = AIAgent.from({
      outputKey: "text",
    });

    spyOn(model, "process")
      .mockReturnValueOnce(Promise.resolve(stringToAgentResponseStream("Hello, ")))
      .mockReturnValueOnce(Promise.resolve(stringToAgentResponseStream("Hello, world!")));

    const team = TeamAgent.from({
      skills: [first, second],
      mode,
    });

    const stream = await aigne.invoke(team, { message: "hello" }, { streaming: true });

    expect(readableStreamToArray(stream)).resolves.toMatchSnapshot();
  },
);

test("TeamAgent with sequential mode should yield output chunks correctly", async () => {
  const teamAgent = TeamAgent.from({
    name: "sequential-team",
    mode: ProcessMode.sequential,
    skills: [
      FunctionAgent.from({
        name: "search",
        process: async ({ question }: Message) => {
          return {
            question,
            result: [
              { title: "First Result", link: "https://example.com/1" },
              { title: "Second Result", link: "https://example.com/2" },
            ],
          };
        },
      }),
      AIAgent.from({
        name: "summarizer",
        instructions: "Summarize the search results:\n{{result}}",
        outputKey: "summary",
      }),
    ],
  });

  const aigne = new AIGNE({ model: new OpenAIChatModel() });

  assert(aigne.model);
  spyOn(aigne.model, "process").mockReturnValueOnce(
    stringToAgentResponseStream("First Result, Second Result"),
  );

  const stream = await aigne.invoke(teamAgent, { question: "What is AIGNE?" }, { streaming: true });
  expect(await readableStreamToArray(stream)).toMatchSnapshot();
});
