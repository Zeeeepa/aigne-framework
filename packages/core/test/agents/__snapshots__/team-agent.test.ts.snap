// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`TeamAgent should return streaming response with sequential process method (multiple agent with different output keys) 1`] = `
[
  {
    "delta": {
      "text": {
        "first": "Hello",
      },
    },
  },
  {
    "delta": {
      "text": {
        "first": ",",
      },
    },
  },
  {
    "delta": {
      "text": {
        "first": " ",
      },
    },
  },
  {
    "delta": {
      "text": {
        "second": "Hello",
      },
    },
  },
  {
    "delta": {
      "text": {
        "second": ",",
      },
    },
  },
  {
    "delta": {
      "text": {
        "second": " ",
      },
    },
  },
  {
    "delta": {
      "text": {
        "second": "world",
      },
    },
  },
  {
    "delta": {
      "text": {
        "second": "!",
      },
    },
  },
]
`;

exports[`TeamAgent should return streaming response with parallel process method (multiple agent with different output keys) 1`] = `
[
  {
    "delta": {
      "text": {
        "first": "Hello",
      },
    },
  },
  {
    "delta": {
      "text": {
        "second": "Hello",
      },
    },
  },
  {
    "delta": {
      "text": {
        "first": ",",
      },
    },
  },
  {
    "delta": {
      "text": {
        "second": ",",
      },
    },
  },
  {
    "delta": {
      "text": {
        "first": " ",
      },
    },
  },
  {
    "delta": {
      "text": {
        "second": " ",
      },
    },
  },
  {
    "delta": {
      "text": {
        "second": "world",
      },
    },
  },
  {
    "delta": {
      "text": {
        "second": "!",
      },
    },
  },
]
`;

exports[`TeamAgent should return streaming response with sequential process method (multiple agent with same output key) 1`] = `
[
  {
    "delta": {
      "text": {
        "text": "Hello",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": ",",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": " ",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": "Hello",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": ",",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": " ",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": "world",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": "!",
      },
    },
  },
]
`;

exports[`TeamAgent should return streaming response with parallel process method (multiple agent with same output key) 1`] = `
[
  {
    "delta": {
      "text": {
        "text": "Hello",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": ",",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": " ",
      },
    },
  },
]
`;

exports[`TeamAgent with sequential mode should yield output chunks correctly 1`] = `
[
  {
    "delta": {
      "json": {
        "question": "What is AIGNE?",
        "result": [
          {
            "link": "https://example.com/1",
            "title": "First Result",
          },
          {
            "link": "https://example.com/2",
            "title": "Second Result",
          },
        ],
      },
    },
  },
  {
    "delta": {
      "text": {
        "summary": "First",
      },
    },
  },
  {
    "delta": {
      "text": {
        "summary": " ",
      },
    },
  },
  {
    "delta": {
      "text": {
        "summary": "Result",
      },
    },
  },
  {
    "delta": {
      "text": {
        "summary": ",",
      },
    },
  },
  {
    "delta": {
      "text": {
        "summary": " ",
      },
    },
  },
  {
    "delta": {
      "text": {
        "summary": "Second",
      },
    },
  },
  {
    "delta": {
      "text": {
        "summary": " ",
      },
    },
  },
  {
    "delta": {
      "text": {
        "summary": "Result",
      },
    },
  },
]
`;
